backends/dpdk/dpdkArch.h

```diff
@@ -446,32 +446,95 @@ class ConvertInternetChecksum : public PassManager {
  * for emitting to the .spec file later */
 class CollectRegisterDeclaration : public Inspector {
     std::map<const IR::Declaration_Instance *, cstring> *reg_map;
-    
+    P4::TypeMap *typeMap;
+
   public:
     CollectRegisterDeclaration(
-        std::map<const IR::Declaration_Instance *, cstring> *reg_map, P4::TypeMap *)
-        : reg_map(reg_map) {}
+        std::map<const IR::Declaration_Instance *, cstring> *reg_map, P4::TypeMap *typeMap)
+        : reg_map(reg_map),  typeMap(typeMap) {}
 
     bool preorder(const IR::Declaration_Instance *d) override {
         if (d->type->is<IR::Type_Specialized>()) {
             auto type = d->type->to<IR::Type_Specialized>();
             auto externTypeName = type->baseType->path->name.name;
             if (externTypeName == "Register"){
-              if (d->arguments->size() != 1 and d->arguments->size() != 2 ) {
-                  ::error("%1%: expected size and optionally init_val as arguments", d);
-              }
-              reg_map->emplace(d, d->name);
+                if (d->arguments->size() != 1 and d->arguments->size() != 2 ) {
+                    ::error("%1%: expected size and optionally init_val as arguments", d);
+                }
+                reg_map->emplace(d, d->name);
+            }
+        }
+        return false;
+    }
+};
+
+/* This pass collects PSA meter declaration instances and push them to a map
+ * for emitting to the .spec file later */
+class CollectMeterDeclaration : public Inspector {
+    std::map<const IR::Declaration_Instance *, cstring> *met_map;
+    P4::TypeMap *typeMap;
+
+  public:
+    CollectMeterDeclaration(
+        std::map<const IR::Declaration_Instance *, cstring> *met_map, P4::TypeMap *typeMap)
+        : met_map(met_map) , typeMap(typeMap) {}
+
+    bool preorder(const IR::Declaration_Instance *d) override {
+        if (d->type->is<IR::Type_Specialized>()) {
+            auto type = d->type->to<IR::Type_Specialized>();
+            auto externTypeName = type->baseType->path->name.name;
+            if (externTypeName == "Meter"){
+                if (d->arguments->size() != 2) {
+                    ::error("%1%: expected number of meters and type of meter as arguments", d);
+                } else {
+                    if (d->arguments->at(1) == 0)
```

Comment:
I am facing an issue w.r.t adding a test for this case.

I added a test in testdata/p4_16_samples/ for emitting this warning and added the warning in reference stderr output "psa-example-dpdk-meter1.p4-stderr". This test compiles as expected and produces a warning when compiled with p4c-dpdk.  However, **make check** fails while compiling this test using _p4test_. It fails because of output mismatch as there is no warning when compiling with p4test and reference stderr output has warning. What should I do in this case?

Reply:
So far the p4 reference outputs have always been produced by p4test.
Is the dpdk backend testing framework also saving reference outputs?
Ideally it should only save the .spec output. Then there should be no conflict: the .p4 and -stderr outputs come from p4test, and the .spec output from p4c-dpdk. 
If both compilers produce reference outputs we have a second problem, since the test has an inherent data race. When we run tests on multiple cores they could overwrite the reference outputs files. So this should not be done. The likelihood of this happening is low, since we have many tests, but it's not zero.
Finally, note that we have already this problem where a test should pass with p4test and fail with a specific backend which has more constraints. Then the solution is to list the test as xfail into the makefile of the backend where it is supposed to fail, with comments explaining why.